_posts/alex_lepsa/2018-03-22-Spring-Method-Security-Using-JWTs.md

```diff
@@ -0,0 +1,342 @@
+        handler.setApplicationContext(context);
+        return handler;
+    }
+
+    @Bean
+    public StudyJWTStrategy studySecurity() {
+        return new StudyJWTStrategy();
+    }
+
+}
+```
+
+Since both the new JWT based method security configuration as well as the legacy configuration both provide the `studySecurity()` method, we can easily switch between the two very easily.
+
+# The Meat (or Tofu) of the Solution #
+
+Now that we have our configs updated to use the new JWT strategy, let's take a look at the `StudyJWTStrategy` class where we define our strategy and the details of how we verify a token.
+
+```
+@Slf4j
+@Component
+@Profile("jwt")
+public class StudyJWTStrategy implements StudyStrategyInterface {
+
+    @Value("${auth.server.prefix}")
+    protected String scopePrefix;
+
+    @Value("${auth.server.suffix}")
+    protected String scopeSuffix;
+
+    public boolean authorize(@NonNull Authentication authentication, @NonNull final String studyId) {
+        log.info("Checking authorization with study id {}", studyId);
+
+        val details = (OAuth2AuthenticationDetails) authentication.getDetails();
+        val user = (JWTUser) details.getDecodedDetails();
+
+        return verify(user, studyId);
+    }
+
+    boolean verify(JWTUser user, String studyId) {
+        final val roles = user.getRoles();
+        val check = roles.stream().filter(s -> isGranted(s, studyId)).collect(toList());
+        return !check.isEmpty();
+    }
+
+    private boolean isGranted(String tokenScope, String studyId) {
+        log.info("Checking JWT's scope '{}', server's scopePrefix='{}', studyId '{}', scopeSuffix='{}'",
+                tokenScope, scopePrefix, studyId, scopeSuffix);
+        return getSystemScope().equals(tokenScope) || getEndUserScope(studyId).equals(tokenScope); //short-circuit
+    }
+
+    private String getEndUserScope(String studyId) {
+        return DOT.join(scopePrefix, studyId.toUpperCase(), scopeSuffix);
+    }
+
+    private String getSystemScope() {
+        return DOT.join(scopePrefix, scopeSuffix);
+    }
+}
+```
+
+There are a few things going on here so let's walk through this file. First we have a couple member variables that are instantiated auto-magically, these dictate the prefix and postfix required in the extracted user role and are configurable in `applcation.yml`, this is our approach from the original OAuth2 scopes strategy that we (at least for now) are porting over to the JWT strategy. Next we have the main authorize method that is what eventually get's called from the `@PreAuthorize` annotation, it extracts the authentication details (which you'll notice is `OAuth2AuthenticationDetails` - more on that later), converts to a `JWTUser` object (more on that too later) and finally returns the result of the verify method. Verify is pretty straightforward, it checks the roles the `JWTUser` possesses against what is constructed as the required role given the studyId, scopePrefix, and scopeSuffix.
```

Comment:
Break this paragraph in multiple bullet points ?

Reply:
I've broken up this paragraph as you're right it was much to dense, however I'm just not a fan of bullets most of the time unless they are very short, for this sort of stuff I prefer full sentences.